[Buildstream SDK] x265 encoder support

philn opened the pull request:
#### ca2221a0959821533dfe3a24f6929284bf6e1f3c
<pre>
[Buildstream SDK] x265 encoder support
<a href="https://bugs.webkit.org/show_bug.cgi?id=259242">https://bugs.webkit.org/show_bug.cgi?id=259242</a>

Reviewed by Carlos Alberto Lopez Perez.

For WPT WebCodec tests we need an HEVC software encoder, since the bots don&apos;t have GPU capabilities
for hardware encoding.

* Tools/buildstream/elements/freedesktop-sdk.bst:
* Tools/buildstream/elements/sdk-platform.bst:
* Tools/buildstream/patches/fdo-0009-gst-plugins-bad-Enable-x265-encoder.patch: Added.

Canonical link: <a href="https://commits.webkit.org/266098@main">https://commits.webkit.org/266098@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a42b3050573d483951229011cbe23432502a54a1

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/12829 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/13155 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/13482 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/14568 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/12248 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/12891 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/15657 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/13175 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/14954 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/12994 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/13723 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/10859 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/15019 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/11011 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/11609 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/18679 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/12086 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/11778 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/14979 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/12232 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/10142 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/11495 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/15808 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/1460 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/12073 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

clopez commented on Tools/buildstream/patches/fdo-0009-gst-plugins-bad-Enable-x265-encoder.patch:
```diff
@@ -0,0 +1,405 @@
++             endif()
++         endif()
++     endif()
++-    if(ARM AND CROSS_COMPILE_ARM)
++-        if(ARM64)
++-            set(ARM_ARGS -fPIC)
++-        else()
++-            set(ARM_ARGS -march=armv6 -mfloat-abi=soft -mfpu=vfp -marm -fPIC)
++-        endif()
++-        message(STATUS "cross compile arm")
++-    elseif(ARM)
++-        if(ARM64)
++-            set(ARM_ARGS -fPIC)
++-            add_definitions(-DHAVE_NEON)
++-        else()
++-            find_package(Neon)
++-            if(CPU_HAS_NEON)
++-                set(ARM_ARGS -mcpu=native -mfloat-abi=hard -mfpu=neon -marm -fPIC)
++-                add_definitions(-DHAVE_NEON)
++-            else()
++-                set(ARM_ARGS -mcpu=native -mfloat-abi=hard -mfpu=vfp -marm)
++-            endif()
++-        endif()
```
This logic doesn't make much sense to me.

1. Why making a difference between the flags passed to the compiler when cross-compiling and when not?. We do care about redistributable binaries on the SDK. So we should always build with compatible flags to the minimum baseline supported at run-time and not use as baseline whatever the CPU that builds the SDK happens to have. So I suggest to avoid using flags like `-mcpu=native`
2. Why use `march=armv6 -mfloat-abi=soft` for ARMv7 when cross-compiling but end using `-mcpu=native -mfloat-abi=hard` when building natively for ARMv7?

philn replied:
I copied those patches from https://github.com/rpmfusion/x265/
I can remove the ones related with armv6/7 if needed, we don't support those legacy arches anyway.

clopez replied:
ok, is fine I guess. If in the end we end supporting armv7 this needs an update but right now we don't support it.

clopez commented on Tools/buildstream/patches/fdo-0009-gst-plugins-bad-Enable-x265-encoder.patch:
```diff
@@ -0,0 +1,405 @@
++             endif()
++         endif()
++     endif()
++-    if(ARM AND CROSS_COMPILE_ARM)
++-        set(ARM_ARGS -march=armv6 -mfloat-abi=soft -mfpu=vfp -marm -fPIC)
++-    elseif(ARM)
++-        find_package(Neon)
++-        if(CPU_HAS_NEON)
++-            set(ARM_ARGS -mcpu=native -mfloat-abi=hard -mfpu=neon -marm -fPIC)
++-            add_definitions(-DHAVE_NEON)
++-        else()
++-            set(ARM_ARGS -mcpu=native -mfloat-abi=hard -mfpu=vfp -marm)
++-        endif()
```
same here than commented above.

philn committed:
[Buildstream SDK] x265 encoder support
https://bugs.webkit.org/show_bug.cgi?id=259242

Reviewed by Carlos Alberto Lopez Perez.

For WPT WebCodec tests we need an HEVC software encoder, since the bots don't have GPU capabilities
for hardware encoding.

* Tools/buildstream/elements/freedesktop-sdk.bst:
* Tools/buildstream/elements/sdk-platform.bst:
* Tools/buildstream/patches/fdo-0009-gst-plugins-bad-Enable-x265-encoder.patch: Added.

Canonical link: https://commits.webkit.org/266098@main
